website/blog/2020-12-04-typedset.md

```diff
@@ -0,0 +1,53 @@
+---
+title: Using OptimizedTypedSet to Improve Map and Array Functions
+author: Ying Su
+authorURL: https://www.linkedin.com/in/ying-su-b00b81107/
+authorFBID: 656599427
+---
+
+Ying Su
+
+Function evaluation is a big part of projection CPU cost. Recently we optimized a set of functions using `TypedSet`, e.g.  `map_concat`, `array_union`, `array_intersect`, and `array_except`. By introducing a new `OptimizedTypeSet`, the above functions were improved for 40% to 80% in CPU and wall time as shown in the JMH benchmarks, and the production queries containing these functions improved for about 5% overall. Meanwhile, the reserved memory and allocation rate dropped for 10% and up to 80% respectively. Furthermore, it resolved the long knowing issue of throwing `EXCEEDED_FUNCTION_MEMORY_LIMIT` for large incoming blocks: "The input to function_name is too large. More than 4MB of memory is needed to hold the intermediate hash set.” 
+
+The `OptimizedTypeSet` and improvements to the above mentioned functions are merged to master, and will be available from Presto 0.244.
+
+<!--truncate-->
+
+In this post we will look at the methods used in this improvement.
+
+## Avoid Using Internal BlockBuilder
+
+`TypedSet` has an internal `BlockBuilder` and appends each `Block` position being added to it, as well as an external `BlockBuilder` to construct the results. Block building using `BlockBuilder` is very costly and inefficient, especially when the memory growth was not handled properly. The `BlockBuilder` was needed to keep track of the elements being added, so that it can:
+
+1. Resolve hash table probing collision 
+2. Rehash. 
+
+However, both are actually not a problem. In all use cases, whole blocks (instead of several positions of a block) are added to the TypedSet, and the problem of resolving collisions can be resolved by keeping track of the blocks being added. Rehashing is not needed at all since we can know the max number of entries before creating a `TypedSet` for most use cases. 
```

Comment:
```suggestion
However, `BlockBuilder` is not necessary for either of these issues. In all use cases, whole blocks (instead of several positions of a block) are added to the TypedSet, and the problem of resolving collisions can be resolved by keeping track of the blocks being added. Rehashing is not needed in most cases because we can calculate the maximum number of entries before creating the `TypedSet`. 
```